add inverse kinematics class by using least square function from scipy

AurelienRenou opened the pull request.

Commit:
class inverse kinematics and example

Commit:
uses of  scipy.optimize.least_squares instead of quad diff, add get_range min and max in utils

Ipuch commented:
related to #2

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+    """
+    The class for generate inverse kinemacs from c3d files
+    """
+
```
docstrings attributes and methods

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+    def __init__(
+            self,
+            biorbd_model_path,
+            c3d_path_file,
```
add hints

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+        self.nb_frames = self.c3d_data.nb_frames
+        self.nb_markers = self.biorbd_model.nbMarkers()
+        self.model_markers = np.zeros((3, self.nb_markers))
+        self.q = np.zeros((self.biorbd_model.nbQ(), self.nb_frames))
```
add a comment that you initialize attributes add a space with the line before

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+        self.list_model_markers = [
+            self.biorbd_model.markerNames()[i].to_string() for i in range(len(self.biorbd_model.markerNames()))
+        ]
+        self.markers_data = self.c3d_data.trajectories
```
try to be consistent, if you use model_markers. Use xp_markers instead of markers_data.

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+    def _marker_diff(self, q, model_xp):
+        """
+        Compute the difference between the marker position in the model and the position in the data
+        """
```
hints and attributes

Commit:
add self.nbQ and clean reshape size of model_xp

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+    def _marker_jac(self, q, model_xp):
+        """
+
+        """
```
hints attriutes

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+        mat_Jac = self.biorbd_model.technicalMarkersJacobian(q)
+
+        Jac = np.zeros((3 * self.nb_markers, self.biorbd_model.nbQ())) # todo:nb_q in self, modify the name
+        for m in range(self.nb_markers):
```
resolve the todo, jac in lower case plz :)

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+
+        return Jac
+
+    def solve(self, method: str = "lm"):
```
attributes and hints

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+        c3d_markers = self.c3d_data.marker_names
+        model_markers = [i.to_string() for i in self.biorbd_model.markerNames()]
+
+        if method == "trf":
+            for ii in range(1, self.nb_frames):
+                print(f" ****   Frame {ii}  ****")
+                x0 = self.q[:, ii - 1]
+                sol = scipy.optimize.least_squares(
+                    fun=self._marker_diff,
+                    args=[self.markers_data[:, :, ii]],
+                    # bounds=(bounds_min, bounds_max),
+                    jac=self._marker_jac,
+                    x0=x0,
+                    method=method,
+                    xtol=1e-6,
+                    tr_options=dict(disp=False),
+                )
+                self.q[:, ii] = sol.x
+
+        else:
+            if method != "lm":
+                print("method = lm")
+            # The first frame use the trf method in order to be sure to respect the bounds
+            x0 = np.random.random((self.biorbd_model.nbQ())) * 0.01
+            sol = scipy.optimize.least_squares(
+                fun=self._marker_diff,
+                args=[self.markers_data[:, :, 0]],
+                bounds=(bounds_min, bounds_max),
+                jac=self._marker_jac,
+                x0=x0,
+                method="trf",
+                xtol=1e-6,
+                tr_options=dict(disp=False),
+            )
+            self.q[:, 0] = sol.x
```
initialization is the same even if trf and lm is set. Please merge

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+                )
+                self.q[:, ii] = sol.x
+
+        else:
+            if method != "lm":
```
if method == "trf":
elif method == "lm":
else:
 raise ValueError(" this method is not implemented please use "trf" or "lm" as argument")

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+            if method != "lm":
+                print("method = lm")
+            # The first frame use the trf method in order to be sure to respect the bounds
+            x0 = np.random.random((self.biorbd_model.nbQ())) * 0.01
```
x0 is the same for each methods put it higher

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+        model_markers = [i.to_string() for i in self.biorbd_model.markerNames()]
+
+        if method == "trf":
+            for ii in range(1, self.nb_frames):
```
it doesn't start from zero ?

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -0,0 +1,121 @@
+        bounds_max = np.squeeze(get_range_max_q(self.biorbd_model))
+        bounds_min = np.squeeze(get_range_min_q(self.biorbd_model))
+        c3d_markers = self.c3d_data.marker_names
+        model_markers = [i.to_string() for i in self.biorbd_model.markerNames()]
```
is it used ? if not remove.

Ipuch reviewed (commented):
nice so far. Have you blacked the code ?

Commit:
blacked

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -44,15 +45,15 @@ def _marker_diff(self, q, model_xp):
 
         # return the sum of the squared differences between the model and the data
-        return vect_pos_markers - np.reshape(model_xp.T, (19*3,))  # todo: 19
+        return vect_pos_markers - np.reshape(model_xp.T, (len(model_xp[0])*3,))  
```
use self.nb_markers instead of len(model_xp). 
Rename model_xp by xp_markers

Commit:
review's correction

Ipuch commented on scripts/utils.py:
```diff
@@ -26,6 +26,50 @@ def get_range_q(biorbd_model: biorbd.Model):
+    Parameters
+    ----------
+    biorbd_model: biorbd.Model
+        The biorbd model
+
+    Returns
+    -------
+    range_q
+        The range of the q for each dof
+    """
+    ranges = np.zeros((biorbd_model.nbQ(), 1))
+    for dof_nb in range(biorbd_model.nbDof()):
+        seg_id, dof_id = get_segment_and_dof_id_from_global_dof(biorbd_model, dof_nb)
+        range_max = biorbd_model.segment(seg_id).QRanges()[dof_id].max()
+        ranges[dof_nb] = range_max
+    return ranges
+
+
+def get_range_min_q(biorbd_model: biorbd.Model):
+    """
+
+
+    Parameters
+    ----------
+    biorbd_model: biorbd.Model
+        The biorbd model
+
+    Returns
+    -------
+    range_q
+        The range of the q for each dof
+    """
+    ranges = np.zeros((biorbd_model.nbQ(), 1))
+    for dof_nb in range(biorbd_model.nbDof()):
+        seg_id, dof_id = get_segment_and_dof_id_from_global_dof(biorbd_model, dof_nb)
+        range_min = biorbd_model.segment(seg_id).QRanges()[dof_id].min()
+        ranges[dof_nb] = range_min
+    return ranges
+
+
```
Please delete this

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -15,8 +15,8 @@ class StaticInverseKinematics:
 
```
Please list the attributes of the object and the method of the object

AurelienRenou replied:
Done

Commit:
modify get range q to get a tuple (bound min, bound max), put bound max and min in arguments of the Class

Commit:
solve method shrank, optimize method add

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -60,7 +60,33 @@ def _marker_jac(self, q: np.ndarray, xp_markers: np.ndarray, n_frame: int):
+    def optimize(self, n_frame: int, method: str, bounds: tuple() = None):
+        print(f" ****   Frame {n_frame}  ****")
+        x0 = self.q[:, n_frame - 1]
+        if bounds is None:
+            sol = scipy.optimize.least_squares(
+                    fun=self._marker_diff,
+                    args=([self.xp_markers[:, :, n_frame]]),
+                    jac=self._marker_jac,
+                    x0=x0,
+                    method=method,
+                    xtol=1e-6,
+                    tr_options=dict(disp=False),
+                )
+        else:
+            sol = scipy.optimize.least_squares(
+                fun=self._marker_diff,
+                args=([self.xp_markers[:, :, n_frame]]),
+                bounds=bounds,
+                jac=self._marker_jac,
+                x0=x0,
+                method=method,
+                xtol=1e-6,
+                tr_options=dict(disp=False),
+            )
```
Are you able to delete this if?

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -60,7 +60,33 @@ def _marker_jac(self, q: np.ndarray, xp_markers: np.ndarray, n_frame: int):
+            )
+        self.q[:, n_frame] = sol.x
+
+    def solve(self, method: str = "lm", full: bool = False):
```
What is the meaning of full = True with method = "trf"? If it has no sense, please create a new tag as a method which is "only_lm".

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -60,7 +60,33 @@ def _marker_jac(self, q: np.ndarray, xp_markers: np.ndarray, n_frame: int):
         return jac
 
-    def solve(self, method: str = "lm"):
+    def optimize(self, n_frame: int, method: str, bounds: tuple() = None):
+        print(f" ****   Frame {n_frame}  ****")
```
please add a docstring and modify the print to display
```python3
 print(f" ****   Frame {n_frame} / {self.nb_frame} ****")```

AurelienRenou replied:
done

Ipuch replied:
@AurelienRenou  This one seems not done

AurelienRenou replied:
yes, now it's done

Ipuch reviewed (commented):
we are getting there !

Commit:
add doc string

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -11,6 +11,49 @@
+    nb_markers: int
+        The number of markers in the model
+    model_markers: np.array
+
```
extra space to be removed

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -11,6 +11,49 @@
+        The list of markers' name
+    xp_markers: np.array
+        The position of the markers from the c3d
+    nbQ: int
```
name it consistently nb_q

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -39,6 +82,12 @@ def __init__(
     def _marker_diff(self, q: np.ndarray, xp_markers: np.ndarray):
         """
         Compute the difference between the marker position in the model and the position in the data
+        Arguments:
```
add a space and replace "Arguments" by "Parameters"

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -51,6 +100,13 @@ def _marker_diff(self, q: np.ndarray, xp_markers: np.ndarray):
     def _marker_jac(self, q: np.ndarray, xp_markers: np.ndarray):
         """
+        Generate the Jacobian matrix for each frame.
+        Arguments:
```
same

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -39,6 +82,12 @@ def __init__(
+        q: np.ndarray
+            The q values
+        xp_markers: np.ndarray
+            The position of the markers from the c3d during a certain frame
```
you forgot to specifiy what it returns

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -51,6 +100,13 @@ def _marker_diff(self, q: np.ndarray, xp_markers: np.ndarray):
+            The q values
+        xp_markers: np.ndarray
+            The position of the markers from the c3d during a certain frame
         """
```
you forgot to specifiy what it returns

AurelienRenou replied:
done

Ipuch reviewed (commented):
some tiny details

Commit:
simplify and shrink solve method, choose between lm, trf and only lm, doc string of solve methode

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -116,60 +117,49 @@ def _marker_jac(self, q: np.ndarray, xp_markers: np.ndarray):
 
+        """
+        bounds = self.bounds if method == "trf" else (-np.inf, np.inf)
+        first_method = "lm" if method == "only_lm" else "trf"
```
rename "initial_method"

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -116,60 +117,49 @@ def _marker_jac(self, q: np.ndarray, xp_markers: np.ndarray):
+                        Usually the most efficient method for small unconstrained problems.
 
+        """
+        bounds = self.bounds if method == "trf" else (-np.inf, np.inf)
```
add a "initial_bounds"
 and gather initial_bounds and initial_method

AurelienRenou replied:
done

Ipuch commented on static_inverse_kinematics.py:
```diff
@@ -116,60 +117,49 @@ def _marker_jac(self, q: np.ndarray, xp_markers: np.ndarray):
+        first_method = "lm" if method == "only_lm" else "trf"
+        method = "lm" if method == "only_lm" else method
+
+        if method == "lm" or method == "trf":
```
reverse the if.
```python
 if method != "lm" and method != "trf" and method != "only_lm":
      raise
 else
 ...
 ```

AurelienRenou replied:
done

Commit:
changed nbQ to nb_q, changed Arguments to Parameters, add what's method returned

Ipuch reviewed (commented):
Tonight this is done!

Commit:
solve method update, add initial_bounds and rename first_method by initial_method, inverse the if condition

Commit:
print frame i / number of frame in solve
